# Data Analysis and Visualisation &mdash; Lab 08

## Problem 3

In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
from setuptools.command.rotate import rotate
from scipy import stats
import numpy as np

plt.style.use("ggplot")
plt.rcParams.update({
    "font.size": 12,
})

### Loading data

In [2]:
df = pd.read_csv("../data/laptop-prices/Laptop+Prices.csv")

### Multiple Linear Regression

Proposed question:

*"Which laptop specifications significantly affect the price, and how?"*

**Explanation:**

We perform an Ordinary Least Squares (OLS) regression using CPU frequency, screen size, weight, RAM, storage, and screen resolution to predict price. This quantifies each feature's contribution while controlling for the others.

In [3]:
import statsmodels.api as sm

X = df[["CPU_freq", "Inches", "Weight", "Ram", "PrimaryStorage", "ScreenW", "ScreenH"]]
y = df["Price_euros"]

X = sm.add_constant(X)

model = sm.OLS(y, X).fit()

model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:            Price_euros   R-squared:                       0.668
Model:                            OLS   Adj. R-squared:                  0.666
Method:                 Least Squares   F-statistic:                     363.8
Date:                Sat, 22 Nov 2025   Prob (F-statistic):          7.56e-298
Time:                        07:40:32   Log-Likelihood:                -9460.2
No. Observations:                1275   AIC:                         1.894e+04
Df Residuals:                    1267   BIC:                         1.898e+04
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
==================================================================================
                     coef    std err          t      P>|t|      [0.025      0.975]
----------------------------------------------------------------------------------
const            120.6943    179.221      0.673      0.501    -230.909     472.298
CPU_freq         284.3182     25.056     11.347      0.000     235.163     333.474
Inches           -71.8246     14.810     -4.850      0.000    -100.879     -42.770
Weight            83.1507     32.516      2.557      0.011      19.360     146.941
Ram               77.1641      2.844     27.134      0.000      71.585      82.743
PrimaryStorage    -0.1900      0.032     -5.875      0.000      -0.253      -0.127
ScreenW            0.0338      0.218      0.155      0.877      -0.393       0.461
ScreenH            0.5938      0.379      1.568      0.117      -0.149       1.337
==============================================================================
Omnibus:                      332.496   Durbin-Watson:                   2.043
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             1792.160
Skew:                           1.101   Prob(JB):                         0.00
Kurtosis:                       8.375   Cond. No.                     3.66e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 3.66e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

Summary:

* CPU frequency and RAM are strong positive drivers of price.
* Larger screens slightly decrease price; heavier laptops slightly increase it.
* Storage has a small negative effect; screen resolution (width & height) is mostly insignificant.

Overall, the model explains ~67% of price variance ($R^2 = 0.688$), confirming that processors and RAM dominate price determination in practice.